<a href="https://colab.research.google.com/github/parreira01/python/blob/master/modulo4_curso_introducao_python_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---

## Manipulação de dados utilizando o Python

Além da facilidade de importar dados utilizando o pandas, o uso de dataframes nos permite trabalhar na manipulação desses dados de maneira muito prática. É possível observar a composição dos dados, filtrar informações (o que em base de dados grandes é essencial), tratar os dados que porventura não foram interpretados corretamente pelo pandas ou que já vieram com erros direto da fonte, analisar os dados e até combinar diferentes bases de dados.

No caso, essas serão as funcionalidados abordadas neste módulo. Mas antes devemos importar tanto os pacotes quanto a base de dados. Serão utilizados os pacotes pandas e numpy e os dados serão do Índice Mineiro de Responsabilidade Social, microdados do ENADE 2013 e os índices de rateio da Lei Robin Hood referentes a julho de 2020.

As duas primeiras bases de dados estão disponibilizadas na pasta **datasets**, a última base de dados será importada diretamente do [site](http://www.fjp.mg.gov.br/robin-hood/index.php/indices/mesatual).



---

### Importando arquivos e pacotes que serão necessários

* Pacotes:
  * Pandas.
  * Numpy.

* Base de dados:
  * Índice Mineiro de Responsabilidade Social.
  * Lei Robin Hood.
  * Microdados do ENADE.

In [ ]:
# importando pacotes
import pandas as pd
import numpy as np

In [ ]:
# importando o arquivo dados consulta
df_imrs = pd.read_csv('/content/DadosConsulta.csv', sep = ';', decimal = ',', encoding = 'latin_1')

# observando o objeto
df_imrs

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
1   3106200  ...                               sim
2   3106705  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
5   3112505  ...                               sim
6   3117876  ...                               sim
7   3118601  ...                               sim
8   3124104  ...                               sim
9   3126000  ...                               não
10  3129806  ...                               sim
11  3130101  ...                               não
12  3132206  ...                               sim
13  3133709  ...                               não
14  3134608  ...                               sim
15  3136652  ...                               sim
16  3137601  ...                               sim
17  3140159  ...                               sim
18  3140704  ...                               não
19  3141108  ...                               sim
20  3144805  ...                               sim
21  3136603  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
24  3154606  ...                               sim
25  3154804  ...                               não
26  3155306  ...                               não
27  3156700  ...                               sim
28  3157807  ...                               sim
29  3162922  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim
33  3171204  ...                               sim

[34 rows x 75 columns]

In [ ]:
# importando o arquivo com os dados do enade
df_enade = pd.read_stata('/content/enade2013edit.dta')

# observando o objeto
df_enade

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0       Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1       Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2       Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3       Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4       Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
...                      ...          ...  ...          ...        ...
196850       Educação Física  IES Privada  ...      Auxílio        Não
196851       Educação Física  IES Privada  ...  Sem auxílio        Sim
196852       Educação Física  IES Privada  ...  Sem auxílio        Sim
196853       Educação Física  IES Privada  ...      Auxílio        Não
196854       Educação Física  IES Privada  ...      Auxílio        Não

[196855 rows x 19 columns]

In [ ]:
# importando os dados das alíquotas da lei robin hood
indices = pd.read_html('http://www.fjp.mg.gov.br/robin-hood/index.php/indices/mesatual', decimal = ',', thousands = '.')

# obtendo o tamanho do objeto
print('Quantidade de dataframes que compõem o objeto: {}'.format(len(indices)))

Quantidade de dataframes que compõem o objeto: 2


In [ ]:
# obtendo o dataframe com as alíquotas
df_indices = indices[1]

# observando o objeto
df_indices

IBGE  SEF  ... Mínimo per capita  Índice de participação
0      10    1  ...                 -                0.038541
1      20    2  ...                 -                0.072048
2      30    3  ...                 -                0.029543
3      40    4  ...                 -                0.017551
4      50    5  ...                 -                0.026952
..    ...  ...  ...               ...                     ...
848  7180  718  ...                 -                0.032209
849  7190  719  ...                 -                0.017340
850  7200  720  ...                 -                0.177399
851  7210  721  ...                 -                0.036130
852  7220  722  ...                 -                0.012312

[853 rows x 23 columns]

---

### Composição dos dados

Antes de apresentar as análises de composição dos dados, é interessante observarmos a diferença entre função e atributo no Python. Ambas podem ser aplicadas sobre os objetos, mas enquanto na maioria das vezes é possível passar parâmetros que alteram o resultado final, como a função `read_csv` que resulta em diferentes dataframes caso o separador seja vírgula ou ponto e vírgula, no segundo é devolvido um atributo inerente ao objeto, sendo assim imutável.

Um atributo muito utilizado na análise de composição dos dados é o atributo **shape**.



In [ ]:
# tamanho do dataframe
df_imrs.shape

(34, 75)

Que devolve uma *tupla*, onde o primeiro valor corresponde ao número de linhas e o segundo valor ao número de colunas do dataframe.

In [ ]:
print('Número de linhas do objeto df_imrs: {}'.format(df_imrs.shape[0]))
print('Número de colunas do objeto df_imrs: {}'.format(df_imrs.shape[1]))

Número de linhas do objeto df_imrs: 34
Número de colunas do objeto df_imrs: 75


Para dataframes, que são objetos de duas dimensões, é mais informativo utilizar o atributo shape do que a função len(). Dado que a última retorna somente o número de linhas.

In [ ]:
# função len
len(df_imrs)

34

Além do shape e do len, o Python possui as seguintes funções e atributos para observar a composição dos dados.

* head(): retorna as n primeiras entradas do dataframe, onde por default n = 5. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html).
* tail(): retorna as n últimas entradas do dataframe, onde por default n = 5. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.tail.html).
* sample(): retorna n entradas aleatórias do dataframe, onde por default n = 1. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).
* value_counts(): retorna a quantidade de cada valor único de uma **série**. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html).
* unique(): retorna os valores únicos de uma série. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.unique.html).
* nunique(): retorna a quantidade de valores únicos de um dataframe. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.nunique.html).
* dtypes: retorna o tipo de cada coluna de um dataframe. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html).
* info(): retorna um sumário dos dados. [Documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html).



In [ ]:
# observando cinco primeiras entradas
df_imrs.head()

ibge7  ... Conselho de Educação em Atividade
0  3105004  ...                               sim
1  3106200  ...                               sim
2  3106705  ...                               sim
3  3109006  ...                               sim
4  3110004  ...                               sim

[5 rows x 75 columns]

In [ ]:
# observando as dez primeiras entradas
df_imrs.head(10)

ibge7  ... Conselho de Educação em Atividade
0  3105004  ...                               sim
1  3106200  ...                               sim
2  3106705  ...                               sim
3  3109006  ...                               sim
4  3110004  ...                               sim
5  3112505  ...                               sim
6  3117876  ...                               sim
7  3118601  ...                               sim
8  3124104  ...                               sim
9  3126000  ...                               não

[10 rows x 75 columns]

In [ ]:
# observando as cinco últimas observações
df_imrs.tail()

ibge7  ... Conselho de Educação em Atividade
29  3162922  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim
33  3171204  ...                               sim

[5 rows x 75 columns]

In [ ]:
# observando as dez últimas observações
df_imrs.tail(10)

ibge7  ... Conselho de Educação em Atividade
24  3154606  ...                               sim
25  3154804  ...                               não
26  3155306  ...                               não
27  3156700  ...                               sim
28  3157807  ...                               sim
29  3162922  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim
33  3171204  ...                               sim

[10 rows x 75 columns]

In [ ]:
# observando uma entrada aleatória
df_imrs.sample()

ibge7  ... Conselho de Educação em Atividade
32  3168309  ...                               sim

[1 rows x 75 columns]

In [ ]:
# observando outra entrada aleatória
df_imrs.sample()

ibge7  ... Conselho de Educação em Atividade
7  3118601  ...                               sim

[1 rows x 75 columns]

In [ ]:
# observando uma entrada aleatória - random_state = 10
df_imrs.sample(random_state=10)

ibge7  ... Conselho de Educação em Atividade
22  3149309  ...                               sim

[1 rows x 75 columns]

In [ ]:
# observando outra entrada aleatória - random_state = 10
df_imrs.sample(random_state=10)

ibge7  ... Conselho de Educação em Atividade
22  3149309  ...                               sim

[1 rows x 75 columns]

In [ ]:
# observando cinco entradas aleatórias
df_imrs.sample(5)

ibge7  ... Conselho de Educação em Atividade
8   3124104  ...                               sim
4   3110004  ...                               sim
6   3117876  ...                               sim
30  3162955  ...                               sim
23  3153905  ...                               sim

[5 rows x 75 columns]

In [ ]:
# observando dez porcento de entradas aleatórias
df_imrs.sample(frac=0.1)

ibge7  ... Conselho de Educação em Atividade
20  3144805  ...                               sim
18  3140704  ...                               não
14  3134608  ...                               sim

[3 rows x 75 columns]

In [ ]:
# obtendo a quantidade de valores de uma determinada coluna (por default não considera NaN)
df_imrs['Existência de Conselho Municipal de Educação'].value_counts()

sim    28
não     3
Name: Existência de Conselho Municipal de Educação, dtype: int64

In [ ]:
# obtendo a quantidade de valores de uma determinada coluna (considerando NaN)
df_imrs['Existência de Conselho Municipal de Educação'].value_counts(dropna = False)

sim    28
não     3
NaN     3
Name: Existência de Conselho Municipal de Educação, dtype: int64

In [ ]:
# obtendo os valores únicos da coluna de cursos
df_enade['co_grupo'].unique()

[Medicina Veterinária, Odontologia, Medicina, Agronomia, Farmácia, ..., Radiologia, Agronegócios, Gestão Hospitalar, Gestão Ambiental, Educação Física]
Length: 17
Categories (17, object): [Medicina Veterinária < Odontologia < Medicina < Agronomia ... Agronegócios <
                          Gestão Hospitalar < Gestão Ambiental < Educação Física]

In [ ]:
# obtendo os valores únicos da coluna de cursos
df_enade['co_grupo'].unique().to_list()

['Medicina Veterinária',
 'Odontologia',
 'Medicina',
 'Agronomia',
 'Farmácia',
 'Enfermagem',
 'Fonoaudiologia',
 'Nutrição',
 'Fisioterapia',
 'Serviço Social',
 'Zootecnia',
 'Biomedicina',
 'Radiologia',
 'Agronegócios',
 'Gestão Hospitalar',
 'Gestão Ambiental',
 'Educação Física']

In [ ]:
# obtendo a quantidade de valores únicos para cada coluna
df_enade.nunique()

co_grupo           17
cd_catadd           2
co_uf_curso        27
co_regiao_curso     5
nu_idade           64
tp_sexod            2
ano_fim_2g         57
tp_inscricao        1
co_rs_s4n           6
co_rs_s5n           6
co_rs_s6n           6
co_rs_s7n           7
co_rs_s8n           6
co_rs_s9n           2
co_rs_s101          2
co_rs_s102          2
co_rs_s103         11
co_rs_s11n          2
co_rs_s20n          2
dtype: int64

In [ ]:
# obtendo a quantidade de valores únicos para cada coluna (visualização em dataframe)
df_enade.nunique().reset_index()

index   0
0          co_grupo  17
1         cd_catadd   2
2       co_uf_curso  27
3   co_regiao_curso   5
4          nu_idade  64
5          tp_sexod   2
6        ano_fim_2g  57
7      tp_inscricao   1
8         co_rs_s4n   6
9         co_rs_s5n   6
10        co_rs_s6n   6
11        co_rs_s7n   7
12        co_rs_s8n   6
13        co_rs_s9n   2
14       co_rs_s101   2
15       co_rs_s102   2
16       co_rs_s103  11
17       co_rs_s11n   2
18       co_rs_s20n   2

In [ ]:
# obtendo os tipos das variáveis
df_indices.dtypes

IBGE                                 int64
SEF                                  int64
Municipio                           object
VAF                                float64
Área Geográfica                    float64
População                          float64
População dos 50 mais populosos     object
Educação                            object
Produção de Alimentos               object
Patrimônio Cultural                 object
Meio Ambiente (A+B+C)               object
Programa Saúde da Família           object
Saúde per capita                    object
Receita Própria                     object
Cota Mínima                        float64
Município Minerador                 object
Recursos hídricos                   object
Penitenciárias                      object
Esportes                            object
Turismo                             object
ICMS solidário                      object
Mínimo per capita                   object
Índice de participação             float64
dtype: obje

In [ ]:
# obtendo as informações das variáveis
df_indices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 853 entries, 0 to 852
Data columns (total 23 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   IBGE                             853 non-null    int64  
 1   SEF                              853 non-null    int64  
 2   Municipio                        853 non-null    object 
 3   VAF                              853 non-null    float64
 4   Área Geográfica                  853 non-null    float64
 5   População                        853 non-null    float64
 6   População dos 50 mais populosos  853 non-null    object 
 7   Educação                         853 non-null    object 
 8   Produção de Alimentos            853 non-null    object 
 9   Patrimônio Cultural              853 non-null    object 
 10  Meio Ambiente (A+B+C)            853 non-null    object 
 11  Programa Saúde da Família        853 non-null    object 
 12  Saúde per capita      

In [ ]:
# obtendo as informações das variáveis
df_imrs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 75 columns):
 #   Column                                                                                             Non-Null Count  Dtype  
---  ------                                                                                             --------------  -----  
 0   ibge7                                                                                              34 non-null     int64  
 1   Município                                                                                          34 non-null     object 
 2   Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)                         34 non-null     float64
 3   Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática)                          34 non-null     float64
 4   Índice de Qualidade da Educação para a 3ª série do EM (matemática)                                 34 non-null     float64
 

#### Pandas Profiling

O pandas profiling é um pacote *relativamente* novo que produz relatórios sobre um dataframe do pandas. O objetivo é ter uma visão mais ampla sobre as informações contidas nos dados.

As informações sobre o pacote podem ser observadas no seguinte [link](https://github.com/pandas-profiling/pandas-profiling).

Como o pacote é novo, ele ainda não compõe a lista de pacotes do [PyPi](https://pypi.org/), então a maneira mais simples de instalar é diretamente do código fonte.

In [ ]:
# instalando o pandas profiling
!pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip -q

     \ 36.9MB 1.1MB/s
     |████████████████████████████████| 71kB 2.1MB/s 
     |████████████████████████████████| 61kB 5.8MB/s 
     |████████████████████████████████| 604kB 17.2MB/s 
     |████████████████████████████████| 3.1MB 24.9MB/s 
     |████████████████████████████████| 71kB 8.8MB/s 
     |████████████████████████████████| 296kB 37.0MB/s 


In [ ]:
# importando a função ProfileReport do pacote pandas_profiling
from pandas_profiling import ProfileReport

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
# criar um objeto ProfileReport
report = ProfileReport(df_indices)

In [ ]:
# criar visualização do report
report.to_notebook_iframe()

In [ ]:
# criar relatório html
report.to_file("report.html")

---

### Filtragem dos dados

Na maioria dos casos, analisar todo um dataframe de uma vez é uma tarefa quase que impossível. Até o próprio pandas evita de mostrar todas as informações de um dataframe de uma vez, com o objetivo de não utilizar tanto a memória.

Sendo assim, é mais interessante observar dados filtrados por um valor específico para um coluna específica ou para ambos de uma só vez. Nesse caso, estamos interessados em fazer consultas ao dataframe e filtrar os dados que mais nos interessam.


In [ ]:
# observando as cinco primeiras entradas do objeto
df_imrs.head()

ibge7  ... Conselho de Educação em Atividade
0  3105004  ...                               sim
1  3106200  ...                               sim
2  3106705  ...                               sim
3  3109006  ...                               sim
4  3110004  ...                               sim

[5 rows x 75 columns]

In [ ]:
# filtrando valores de Belo Horizonte
df_imrs['Município' == 'Belo Horizonte']

KeyError: ignored

In [ ]:
# filtrando valores de Belo Horizonte
df_imrs[df_imrs['Município'] == 'Belo Horizonte']

ibge7  ... Conselho de Educação em Atividade
1  3106200  ...                               sim

[1 rows x 75 columns]

In [ ]:
# filtrando valores de Belo Horizonte e Ribeirão das Neves (OR)
df_imrs[(df_imrs['Município'] == 'Belo Horizonte') | (df_imrs['Município'] == 'Ribeirão das Neves')]

ibge7  ... Conselho de Educação em Atividade
1   3106200  ...                               sim
24  3154606  ...                               sim

[2 rows x 75 columns]

In [ ]:
# selecionando os municípios
municipios = ['Belo Horizonte', 'Ribeirão das Neves']

# utilizando a função isin
df_imrs[df_imrs['Município'].isin(municipios)]

ibge7  ... Conselho de Educação em Atividade
1   3106200  ...                               sim
24  3154606  ...                               sim

[2 rows x 75 columns]

In [ ]:
# filtrando valores de Belo Horizonte e Ribeirão das Neves (AND)
df_imrs[(df_imrs['Município'] == 'Belo Horizonte') & (df_imrs['Município'] == 'Ribeirão das Neves')]

Empty DataFrame
Columns: [ibge7, Município, Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática), Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática), Índice de Qualidade da Educação para a 3ª série do EM (matemática), Índice de Qualidade da Educação para a 4ª série /5º ano do EF (Língua Portuguesa), Índice de Qualidade da Educação para a 8ª série/9º ano do EF (Língua Portuguesa), Índice de Qualidade da Educação para a 3ª série do EM (Língua Portuguesa), Índice de Qualidade da Educação para a 4ª série /5º ano do EF          , Índice de Qualidade da Educação para a 8ª série/9º ano  do EF, Índice de Qualidade da Educação para a 3ª série do EM, Índice de Qualidade Geral da Educação, Taxa de alunos da 8ª série/9º ano do fundamental acima do nível BÁSICO - Matemática, Taxa de alunos da 8ª série/9º ano do fundamental acima do nível BÁSICO - Português, Taxa de alunos da 3ª série do médio acima do nível BÁSICO - Matemática, Taxa de alunos da 3ª Série do médio acima do nível BÁSICO - Português, Número de alunos da 8ª série/9º ano do fundamental que fizeram a prova de Matemática do PROEB, Número de alunos da 8ª série/9º ano do fundamental que fizeram a prova de Português do PROEB, Número de alunos da 3ª série do médio que fizeram a prova de Matemática do PROEB, Número de alunos da 3ª série do médio que fizeram a prova de Português do PROEB, Percentual de alunos da 8ª série/9º ano do fundamental que fizeram a prova de Matemática do PROEB, Percentual de alunos da 8ª série/9º ano do fundamental que fizeram a prova de Português do PROEB, Percentual de alunos da 3ª série do médio que fizeram a prova de Matemática do PROEB, Percentual de alunos da 3ª série do médio que fizeram a prova de Português do PROEB, Taxa de alunos em escolas com laboratório de informática - fundamental, Taxa de alunos em escolas com Internet  - fundamental                   , Taxa de alunos em escolas com laboratório de informática - médio, Taxa de alunos em escolas com Internet  - médio, Percentual de Funções Docentes com curso superior_EI, Percentual de Funções Docentes com curso superior_EF, Percentual de Funções Docentes com curso superior_EFAI, Percentual de Funções Docentes com curso superior_EFAF, Percentual de Funções Docentes com curso superior_EM, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 1)_EI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 2)_EI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 3)_EI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 4)_EI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 5)_EI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 1)_EF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 2)_EF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 3)_EF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 4)_EF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 5)_EF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 1)_EFAI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 2)_EFAI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 3)_EFAI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 4)_EFAI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 5)_EFAI, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 1)_EFAF, Percentual de docentes por grupo do indicador de adequação da formação do docente (Grupo 2)_EFAF, Percentual de docentes por grupo do indicador de adequ

In [ ]:
# filtrando valores pela qualidade da educação
df_imrs[df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)'] > 0.5]

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
6   3117876  ...                               sim
9   3126000  ...                               não
11  3130101  ...                               não
12  3132206  ...                               sim
13  3133709  ...                               não
14  3134608  ...                               sim
15  3136652  ...                               sim
18  3140704  ...                               não
19  3141108  ...                               sim
20  3144805  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
25  3154804  ...                               não
26  3155306  ...                               não
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim

[20 rows x 75 columns]

In [ ]:
# filtrando valores por comparação entre as variáveis
df_imrs[df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)'] >
        df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (Língua Portuguesa)']]

ibge7  ... Conselho de Educação em Atividade
9   3126000  ...                               não
14  3134608  ...                               sim
23  3153905  ...                               sim

[3 rows x 75 columns]

In [ ]:
# filtrando valores pela qualidade da educação e que tenha um conselho municipal de educação
df_imrs[(df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)'] > 0.5) &
        (df_imrs['Existência de Conselho Municipal de Educação'] == 'sim')]

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
6   3117876  ...                               sim
12  3132206  ...                               sim
14  3134608  ...                               sim
15  3136652  ...                               sim
19  3141108  ...                               sim
20  3144805  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim

[14 rows x 75 columns]

In [ ]:
# filtrando somente as colunas que são importantes
colunas = ['ibge7', 'Município','Índice de Qualidade Geral da Educação','Existência de Conselho Municipal de Educação','Conselho de Educação em Atividade']

df_imrs[colunas]

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
1   3106200  ...                               sim
2   3106705  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
5   3112505  ...                               sim
6   3117876  ...                               sim
7   3118601  ...                               sim
8   3124104  ...                               sim
9   3126000  ...                               não
10  3129806  ...                               sim
11  3130101  ...                               não
12  3132206  ...                               sim
13  3133709  ...                               não
14  3134608  ...                               sim
15  3136652  ...                               sim
16  3137601  ...                               sim
17  3140159  ...                               sim
18  3140704  ...                               não
19  3141108  ...                               sim
20  3144805  ...                               sim
21  3136603  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
24  3154606  ...                               sim
25  3154804  ...                               não
26  3155306  ...                               não
27  3156700  ...                               sim
28  3157807  ...                               sim
29  3162922  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim
33  3171204  ...                               sim

[34 rows x 5 columns]

In [ ]:
# criando um novo objeto
df_imrs_simples = df_imrs[colunas]
df_imrs_simples

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
1   3106200  ...                               sim
2   3106705  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
5   3112505  ...                               sim
6   3117876  ...                               sim
7   3118601  ...                               sim
8   3124104  ...                               sim
9   3126000  ...                               não
10  3129806  ...                               sim
11  3130101  ...                               não
12  3132206  ...                               sim
13  3133709  ...                               não
14  3134608  ...                               sim
15  3136652  ...                               sim
16  3137601  ...                               sim
17  3140159  ...                               sim
18  3140704  ...                               não
19  3141108  ...                               sim
20  3144805  ...                               sim
21  3136603  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
24  3154606  ...                               sim
25  3154804  ...                               não
26  3155306  ...                               não
27  3156700  ...                               sim
28  3157807  ...                               sim
29  3162922  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim
33  3171204  ...                               sim

[34 rows x 5 columns]

In [ ]:
# misturando as consultas
df_imrs_simples[(df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)'] > 0.5) &
        (df_imrs['Existência de Conselho Municipal de Educação'] == 'sim')]

ibge7  ... Conselho de Educação em Atividade
0   3105004  ...                               sim
3   3109006  ...                               sim
4   3110004  ...                               sim
6   3117876  ...                               sim
12  3132206  ...                               sim
14  3134608  ...                               sim
15  3136652  ...                               sim
19  3141108  ...                               sim
20  3144805  ...                               sim
22  3149309  ...                               sim
23  3153905  ...                               sim
30  3162955  ...                               sim
31  3165537  ...                               sim
32  3168309  ...                               sim

[14 rows x 5 columns]

In [ ]:
# entendo o que está por trás da consulta
[(df_imrs['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)'] > 0.5) &
        (df_imrs['Existência de Conselho Municipal de Educação'] == 'sim')]

[0      True
 1     False
 2     False
 3      True
 4      True
 5     False
 6      True
 7     False
 8     False
 9     False
 10    False
 11    False
 12     True
 13    False
 14     True
 15     True
 16    False
 17    False
 18    False
 19     True
 20     True
 21    False
 22     True
 23     True
 24    False
 25    False
 26    False
 27    False
 28    False
 29    False
 30     True
 31     True
 32     True
 33    False
 dtype: bool]

Nesse ponto é interessante observar que as consultas estão baseadas nas análises dos índices. Com isso, é indicado que o índice apresente valores únicos.

In [ ]:
# utilizando o loc para obter a primeira linha
df_imrs.loc[0]

ibge7                                                                         3105004
Município                                                                      Baldim
Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)       0.68
Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática)        0.33
Índice de Qualidade da Educação para a 3ª série do EM (matemática)               0.25
                                                                               ...   
% de matrícula em tempo integral_EFAI                                               0
% de matrícula em tempo integral_EFAF                                           22.44
% de matrícula em tempo integral_EM                                                 0
Existência de Conselho Municipal de Educação                                      sim
Conselho de Educação em Atividade                                                 sim
Name: 0, Length: 75, dtype: object

In [ ]:
# utilizando o loc para obter a primera linha e a segunda coluna
df_imrs.loc[0, 2]

TypeError: ignored

In [ ]:
# utilizando o loc para obter a primera linha e a segunda coluna
df_imrs.loc[0, 'Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)']

0.68

In [ ]:
# alterando o index
df_imrs.set_index('Município')

ibge7  ...  Conselho de Educação em Atividade
Município                      ...                                   
Baldim                3105004  ...                                sim
Belo Horizonte        3106200  ...                                sim
Betim                 3106705  ...                                sim
Brumadinho            3109006  ...                                sim
Caeté                 3110004  ...                                sim
Capim Branco          3112505  ...                                sim
Confins               3117876  ...                                sim
Contagem              3118601  ...                                sim
Esmeraldas            3124104  ...                                sim
Florestal             3126000  ...                                não
Ibirité               3129806  ...                                sim
Igarapé               3130101  ...                                não
Itaguara              3132206  ...                                sim
Itatiaiuçu            3133709  ...                                não
Jaboticatubas         3134608  ...                                sim
Juatuba               3136652  ...                                sim
Lagoa Santa           3137601  ...                                sim
Mário Campos          3140159  ...                                sim
Mateus Leme           3140704  ...                                não
Matozinhos            3141108  ...                                sim
Nova Lima             3144805  ...                                sim
Nova União            3136603  ...                                sim
Pedro Leopoldo        3149309  ...                                sim
Raposos               3153905  ...                                sim
Ribeirão das Neves    3154606  ...                                sim
Rio Acima             3154804  ...                                não
Rio Manso             3155306  ...                                não
Sabará                3156700  ...                                sim
Santa Luzia           3157807  ...                                sim
São Joaquim de Bicas  3162922  ...                                sim
São José da Lapa      3162955  ...                                sim
Sarzedo               3165537  ...                                sim
Taquaraçu de Minas    3168309  ...                                sim
Vespasiano            3171204  ...                                sim

[34 rows x 74 columns]

In [ ]:
# observando o objeto
df_imrs.head()

ibge7  ... Conselho de Educação em Atividade
0  3105004  ...                               sim
1  3106200  ...                               sim
2  3106705  ...                               sim
3  3109006  ...                               sim
4  3110004  ...                               sim

[5 rows x 75 columns]

In [ ]:
# alterando o index (utilizando o inplace)
df_imrs.set_index('Município', inplace = True)

In [ ]:
# observando o objeto
df_imrs.head()

ibge7  ...  Conselho de Educação em Atividade
Município                ...                                   
Baldim          3105004  ...                                sim
Belo Horizonte  3106200  ...                                sim
Betim           3106705  ...                                sim
Brumadinho      3109006  ...                                sim
Caeté           3110004  ...                                sim

[5 rows x 74 columns]

In [ ]:
# utilizando o loc
df_imrs.loc['Belo Horizonte', 'Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)']

0.49

In [ ]:
# obtendo dois índices
df_imrs.loc[['Belo Horizonte','Caeté']]

ibge7  ...  Conselho de Educação em Atividade
Município                ...                                   
Belo Horizonte  3106200  ...                                sim
Caeté           3110004  ...                                sim

[2 rows x 74 columns]

In [ ]:
# obtendo os dados entre os dois índices
df_imrs.loc['Belo Horizonte':'Caeté']

ibge7  ...  Conselho de Educação em Atividade
Município                ...                                   
Belo Horizonte  3106200  ...                                sim
Betim           3106705  ...                                sim
Brumadinho      3109006  ...                                sim
Caeté           3110004  ...                                sim

[4 rows x 74 columns]

In [ ]:
# utilizando o iloc
df_imrs.iloc[0,1]

0.68

In [ ]:
# resetando o index
df_imrs.reset_index(inplace = True)

In [ ]:
# observando o objeto
df_imrs.head()

Município  ...  Conselho de Educação em Atividade
0          Baldim  ...                                sim
1  Belo Horizonte  ...                                sim
2           Betim  ...                                sim
3      Brumadinho  ...                                sim
4           Caeté  ...                                sim

[5 rows x 75 columns]

---

### Tratamento de dados

Com os dados inseridos em um dataframe do pandas é possível também fazer alterações nesses dados, seja alterando caracteres, transformando o tipo das variáveis ou incluindo/excluindo colunas e linhas. O objetivo é deixar o dataframe corretamente especificado.


#### Substituindo valores com replace e transformando o tipo das colunas

In [ ]:
# cinco primeiras entrada do objeto df_aliquotas
df_indices.head()

IBGE  SEF  ... Mínimo per capita  Índice de participação
0    10    1  ...                 -                0.038541
1    20    2  ...                 -                0.072048
2    30    3  ...                 -                0.029543
3    40    4  ...                 -                0.017551
4    50    5  ...                 -                0.026952

[5 rows x 23 columns]

In [ ]:
# obtendo os tipos das variáveis
df_indices.dtypes

IBGE                                 int64
SEF                                  int64
Municipio                           object
VAF                                float64
Área Geográfica                    float64
População                          float64
População dos 50 mais populosos     object
Educação                            object
Produção de Alimentos               object
Patrimônio Cultural                 object
Meio Ambiente (A+B+C)               object
Programa Saúde da Família           object
Saúde per capita                    object
Receita Própria                     object
Cota Mínima                        float64
Município Minerador                 object
Recursos hídricos                   object
Penitenciárias                      object
Esportes                            object
Turismo                             object
ICMS solidário                      object
Mínimo per capita                   object
Índice de participação             float64
dtype: obje

In [ ]:
# substituindo o caracter ' - ' por valores ausentes
df_indices.replace('-',np.nan, inplace = True)

In [ ]:
# observando o objeto
df_indices.head()

IBGE  SEF  ... Mínimo per capita  Índice de participação
0    10    1  ...               NaN                0.038541
1    20    2  ...               NaN                0.072048
2    30    3  ...               NaN                0.029543
3    40    4  ...               NaN                0.017551
4    50    5  ...               NaN                0.026952

[5 rows x 23 columns]

In [ ]:
# obtendo os tipos das variáveis
df_indices.dtypes

IBGE                                 int64
SEF                                  int64
Municipio                           object
VAF                                float64
Área Geográfica                    float64
População                          float64
População dos 50 mais populosos     object
Educação                            object
Produção de Alimentos               object
Patrimônio Cultural                 object
Meio Ambiente (A+B+C)               object
Programa Saúde da Família           object
Saúde per capita                    object
Receita Própria                     object
Cota Mínima                        float64
Município Minerador                 object
Recursos hídricos                   object
Penitenciárias                      object
Esportes                            object
Turismo                             object
ICMS solidário                      object
Mínimo per capita                   object
Índice de participação             float64
dtype: obje

In [ ]:
# obtendo as colunas que são objeto
df_indices_objeto = df_indices.select_dtypes(include='object')
df_indices_objeto

Municipio  ... Mínimo per capita
0       Abadia dos Dourados  ...               NaN
1                    Abaeté  ...               NaN
2                Abre Campo  ...               NaN
3                   Acaiaca  ...               NaN
4                   Açucena  ...               NaN
..                      ...  ...               ...
848            Virginópolis  ...               NaN
849             Virgolândia  ...               NaN
850  Visconde do Rio Branco  ...               NaN
851            Volta Grande  ...               NaN
852          Wenceslau Braz  ...               NaN

[853 rows x 16 columns]

In [ ]:
# obtendo o nome das colunas
nome_indices_objeto = list(df_indices_objeto.columns.values)
nome_indices_objeto

['Municipio',
 'População dos 50 mais populosos',
 'Educação',
 'Produção de Alimentos',
 'Patrimônio Cultural',
 'Meio Ambiente (A+B+C)',
 'Programa Saúde da Família',
 'Saúde per capita',
 'Receita Própria',
 'Município Minerador',
 'Recursos hídricos',
 'Penitenciárias',
 'Esportes',
 'Turismo',
 'ICMS solidário',
 'Mínimo per capita']

In [ ]:
# excluindo a coluna Município
nome_indices_objeto.remove('Municipio')
nome_indices_objeto

['População dos 50 mais populosos',
 'Educação',
 'Produção de Alimentos',
 'Patrimônio Cultural',
 'Meio Ambiente (A+B+C)',
 'Programa Saúde da Família',
 'Saúde per capita',
 'Receita Própria',
 'Município Minerador',
 'Recursos hídricos',
 'Penitenciárias',
 'Esportes',
 'Turismo',
 'ICMS solidário',
 'Mínimo per capita']

In [ ]:
# substituindo as colunas objeto por float
df_indices[nome_indices_objeto] = df_indices[nome_indices_objeto].astype('float64')

# observando os cinco primeiros valores
df_indices.head()

IBGE  SEF  ... Mínimo per capita  Índice de participação
0    10    1  ...               NaN                0.038541
1    20    2  ...               NaN                0.072048
2    30    3  ...               NaN                0.029543
3    40    4  ...               NaN                0.017551
4    50    5  ...               NaN                0.026952

[5 rows x 23 columns]

In [ ]:
# obtendo os tipos das variáveis
df_indices.dtypes

IBGE                                 int64
SEF                                  int64
Municipio                           object
VAF                                float64
Área Geográfica                    float64
População                          float64
População dos 50 mais populosos    float64
Educação                           float64
Produção de Alimentos              float64
Patrimônio Cultural                float64
Meio Ambiente (A+B+C)              float64
Programa Saúde da Família          float64
Saúde per capita                   float64
Receita Própria                    float64
Cota Mínima                        float64
Município Minerador                float64
Recursos hídricos                  float64
Penitenciárias                     float64
Esportes                           float64
Turismo                            float64
ICMS solidário                     float64
Mínimo per capita                  float64
Índice de participação             float64
dtype: obje

#### Criando novas colunas

In [ ]:
# observando o objeto df_imrs
df_imrs.head()

Município  ...  Conselho de Educação em Atividade
0          Baldim  ...                                sim
1  Belo Horizonte  ...                                sim
2           Betim  ...                                sim
3      Brumadinho  ...                                sim
4           Caeté  ...                                sim

[5 rows x 75 columns]

In [ ]:
# criando uma lista com as variáveis de matemática
matematica = ['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)',
              'Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática)',
              'Índice de Qualidade da Educação para a 3ª série do EM (matemática)']

# somando as notas de matemática
media_matematica = (df_imrs[matematica].sum(axis=1))/3
media_matematica

0     0.420000
1     0.420000
2     0.420000
3     0.466667
4     0.436667
5     0.360000
6     0.386667
7     0.393333
8     0.366667
9     0.560000
10    0.370000
11    0.423333
12    0.440000
13    0.476667
14    0.483333
15    0.470000
16    0.413333
17    0.350000
18    0.456667
19    0.413333
20    0.470000
21    0.366667
22    0.406667
23    0.443333
24    0.340000
25    0.400000
26    0.466667
27    0.350000
28    0.390000
29    0.433333
30    0.466667
31    0.433333
32    0.363333
33    0.400000
dtype: float64

In [ ]:
# primeira forma de criar uma nova coluna
df_imrs['Media matemática'] = media_matematica
df_imrs.head()

Município    ibge7  ...  Conselho de Educação em Atividade  Media matemática
0          Baldim  3105004  ...                                sim          0.420000
1  Belo Horizonte  3106200  ...                                sim          0.420000
2           Betim  3106705  ...                                sim          0.420000
3      Brumadinho  3109006  ...                                sim          0.466667
4           Caeté  3110004  ...                                sim          0.436667

[5 rows x 76 columns]

In [ ]:
# segunda forma de criar uma nova coluna
df_imrs.insert(5, 'Media_matematica', media_matematica)
df_imrs.head()

Município    ibge7  ...  Conselho de Educação em Atividade  Media matemática
0          Baldim  3105004  ...                                sim          0.420000
1  Belo Horizonte  3106200  ...                                sim          0.420000
2           Betim  3106705  ...                                sim          0.420000
3      Brumadinho  3109006  ...                                sim          0.466667
4           Caeté  3110004  ...                                sim          0.436667

[5 rows x 77 columns]

In [ ]:
# alterando a coluna Media matemática
df_imrs['Media matemática'].where(df_imrs['Media matemática'] > 0.45, 0.0, inplace = True)
df_imrs.head()

Município    ibge7  ...  Conselho de Educação em Atividade  Media matemática
0          Baldim  3105004  ...                                sim          0.000000
1  Belo Horizonte  3106200  ...                                sim          0.000000
2           Betim  3106705  ...                                sim          0.000000
3      Brumadinho  3109006  ...                                sim          0.466667
4           Caeté  3110004  ...                                sim          0.000000

[5 rows x 77 columns]

#### Incluindo linha

In [ ]:
# observando o objeto com os dados do enade
df_enade.head()

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0  Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim

[5 rows x 19 columns]

In [ ]:
# criando nova linha
nova_linha = {'ano_fim_2g': 2010,
 'cd_catadd': 'IES Privada',
 'co_grupo': 'Educação Física',
 'co_regiao_curso': 'Sudeste',
 'co_rs_s101': 'Prouni integral',
 'co_rs_s102': 'Não Prouni parcial',
 'co_rs_s103': 'Curso pago',
 'co_rs_s11n': 'Auxílio',
 'co_rs_s20n': 'Não',
 'co_rs_s4n': 'Ensino Médio',
 'co_rs_s5n': 'Ensino Médio',
 'co_rs_s6n': 'Outras pessoas',
 'co_rs_s7n': 'De 3 a 4,5 salários mínimos',
 'co_rs_s8n': 'Tenho renda/contribuo com o sustento',
 'co_rs_s9n': 'Não trabalho',
 'co_uf_curso': 'Minas Gerais',
 'nu_idade': 27,
 'tp_inscricao': 'Concluinte',
 'tp_sexod': 'Homem'}

In [ ]:
# incluindo a nova linha no dataframe
df_enade = df_enade.append(nova_linha, ignore_index=True)

# observando a última linha do dataframe
df_enade.tail(1)

co_grupo    cd_catadd  ... co_rs_s11n co_rs_s20n
196855  Educação Física  IES Privada  ...    Auxílio        Não

[1 rows x 19 columns]

#### Alterando nomes nas colunas e nas linhas

In [ ]:
# observando o objeto com os dados do enade
df_enade.head()

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0  Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim

[5 rows x 19 columns]

In [ ]:
# alterando o nome da coluna tp_sexod
df_enade.rename(columns = {'tp_sexod' : 'genero'}, inplace = True)
df_enade.head()

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0  Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim

[5 rows x 19 columns]

In [ ]:
# alterando o nome das linhas (séries apenas)
df_enade['genero'] = df_enade['genero'].map({'Homem' : 0, 'Mulher' : 1})
df_enade.head()

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0  Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim

[5 rows x 19 columns]

#### Organizando os dados

In [ ]:
# alterando o index (utilizando o inplace)
df_imrs.set_index('Município', inplace = True)
df_imrs.head()

ibge7  ...  Media matemática
Município                ...                  
Baldim          3105004  ...          0.000000
Belo Horizonte  3106200  ...          0.000000
Betim           3106705  ...          0.000000
Brumadinho      3109006  ...          0.466667
Caeté           3110004  ...          0.000000

[5 rows x 76 columns]

In [ ]:
# organizando os valores pelo índice qualidade geral da educação
df_imrs.sort_values('Índice de Qualidade Geral da Educação',inplace = True)
df_imrs.head()

ibge7  ...  Media matemática
Município                    ...                  
Mário Campos        3140159  ...               0.0
Ribeirão das Neves  3154606  ...               0.0
Confins             3117876  ...               0.0
Sabará              3156700  ...               0.0
Capim Branco        3112505  ...               0.0

[5 rows x 76 columns]

In [ ]:
# observando os últimos valores
df_imrs.tail()

ibge7  ...  Media matemática
Município             ...                  
Mateus Leme  3140704  ...          0.456667
Juatuba      3136652  ...          0.470000
Nova Lima    3144805  ...          0.470000
Florestal    3126000  ...          0.560000
Rio Manso    3155306  ...          0.466667

[5 rows x 76 columns]

In [ ]:
# organizando os valores pelo índice qualidade geral da educação (do maior para o menor)
df_imrs.sort_values('Índice de Qualidade Geral da Educação',inplace = True, ascending = False)
df_imrs.head()

ibge7  ...  Media matemática
Município             ...                  
Rio Manso    3155306  ...          0.466667
Florestal    3126000  ...          0.560000
Nova Lima    3144805  ...          0.470000
Juatuba      3136652  ...          0.470000
Mateus Leme  3140704  ...          0.456667

[5 rows x 76 columns]

In [ ]:
# organizando os valores pelo índice qualidade geral da educação (do maior para o menor) e utilizando outra variável como "desempate"
df_imrs.sort_values(by=['Índice de Qualidade Geral da Educação','Media_matematica'],inplace = True, ascending = False)
df_imrs.head()

ibge7  ...  Media matemática
Município             ...                  
Florestal    3126000  ...          0.560000
Rio Manso    3155306  ...          0.466667
Nova Lima    3144805  ...          0.470000
Juatuba      3136652  ...          0.470000
Mateus Leme  3140704  ...          0.456667

[5 rows x 76 columns]

#### Excluindo linhas e colunas com drop

In [ ]:
# excluindo coluna com o código do ibge
df_imrs.drop(labels = 'ibge7', axis = 1, inplace = True)
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Media matemática
Município                                                                                ...                  
Florestal                                                 0.73                           ...          0.560000
Rio Manso                                                 0.58                           ...          0.466667
Nova Lima                                                 0.62                           ...          0.470000
Juatuba                                                   0.56                           ...          0.470000
Mateus Leme                                               0.54                           ...          0.456667

[5 rows x 75 columns]

In [ ]:
# excluindo duas colunas
df_imrs.drop(labels = ['Media matemática', 'Media_matematica'], axis = 1, inplace = True)
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Conselho de Educação em Atividade
Município                                                                                ...                                   
Florestal                                                 0.73                           ...                                não
Rio Manso                                                 0.58                           ...                                não
Nova Lima                                                 0.62                           ...                                sim
Juatuba                                                   0.56                           ...                                sim
Mateus Leme                                               0.54                           ...                                não

[5 rows x 73 columns]

In [ ]:
# excluindo linhas
df_imrs.drop(labels = 'Rio Manso', axis = 0, inplace = True)
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Conselho de Educação em Atividade
Município                                                                                ...                                   
Florestal                                                 0.73                           ...                                não
Nova Lima                                                 0.62                           ...                                sim
Juatuba                                                   0.56                           ...                                sim
Mateus Leme                                               0.54                           ...                                não
Itaguara                                                  0.58                           ...                                sim

[5 rows x 73 columns]

In [ ]:
# excluindo duas linhas
df_imrs.drop(labels = ['Mateus Leme','Itaguara'], axis = 0, inplace = True)
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Conselho de Educação em Atividade
Município                                                                                  ...                                   
Florestal                                                   0.73                           ...                                não
Nova Lima                                                   0.62                           ...                                sim
Juatuba                                                     0.56                           ...                                sim
Jaboticatubas                                               0.69                           ...                                sim
Brumadinho                                                  0.61                           ...                                sim

[5 rows x 73 columns]

#### Valores ausentes

In [ ]:
# observando o objeto df_aliquotas
df_indices.head()

IBGE  SEF  ... Mínimo per capita  Índice de participação
0    10    1  ...               NaN                0.038541
1    20    2  ...               NaN                0.072048
2    30    3  ...               NaN                0.029543
3    40    4  ...               NaN                0.017551
4    50    5  ...               NaN                0.026952

[5 rows x 23 columns]

In [ ]:
# observando os valores ausentes
df_indices.isna()

IBGE    SEF  ...  Mínimo per capita  Índice de participação
0    False  False  ...               True                   False
1    False  False  ...               True                   False
2    False  False  ...               True                   False
3    False  False  ...               True                   False
4    False  False  ...               True                   False
..     ...    ...  ...                ...                     ...
848  False  False  ...               True                   False
849  False  False  ...               True                   False
850  False  False  ...               True                   False
851  False  False  ...               True                   False
852  False  False  ...               True                   False

[853 rows x 23 columns]

In [ ]:
# observando a quantidade de valores ausentes
df_indices.isna().sum()

IBGE                                 0
SEF                                  0
Municipio                            0
VAF                                  0
Área Geográfica                      0
População                            0
População dos 50 mais populosos    803
Educação                           229
Produção de Alimentos                1
Patrimônio Cultural                 50
Meio Ambiente (A+B+C)              269
Programa Saúde da Família          457
Saúde per capita                     2
Receita Própria                      2
Cota Mínima                          0
Município Minerador                480
Recursos hídricos                  743
Penitenciárias                     667
Esportes                           445
Turismo                            511
ICMS solidário                      56
Mínimo per capita                  845
Índice de participação               0
dtype: int64

In [ ]:
# observando a quantidade de valores ausentes
df_indices.isna().sum().sort_values(ascending = False)

Mínimo per capita                  845
População dos 50 mais populosos    803
Recursos hídricos                  743
Penitenciárias                     667
Turismo                            511
Município Minerador                480
Programa Saúde da Família          457
Esportes                           445
Meio Ambiente (A+B+C)              269
Educação                           229
ICMS solidário                      56
Patrimônio Cultural                 50
Receita Própria                      2
Saúde per capita                     2
Produção de Alimentos                1
VAF                                  0
SEF                                  0
Municipio                            0
Índice de participação               0
Área Geográfica                      0
População                            0
Cota Mínima                          0
IBGE                                 0
dtype: int64

In [ ]:
# 80% dos valores
print(round(df_indices.shape[0]*0.8))

682


In [ ]:
# dropando as variáveis
df_indices.dropna(axis = 1, thresh=round(df_indices.shape[0]*0.8), inplace = True)
df_indices.head()

IBGE  SEF  ... ICMS solidário  Índice de participação
0    10    1  ...       0.037007                0.038541
1    20    2  ...       0.123040                0.072048
2    30    3  ...       0.071239                0.029543
3    40    4  ...       0.021148                0.017551
4    50    5  ...       0.050144                0.026952

[5 rows x 13 columns]

In [ ]:
# observando a quantidade de valores ausentes
df_indices.isna().sum().sort_values(ascending = False)

ICMS solidário            56
Patrimônio Cultural       50
Receita Própria            2
Saúde per capita           2
Produção de Alimentos      1
Índice de participação     0
Cota Mínima                0
População                  0
Área Geográfica            0
VAF                        0
Municipio                  0
SEF                        0
IBGE                       0
dtype: int64

In [ ]:
# dropando os valores ausentes das linhas com valores ausentes do ICMS solidário e patrimônio cultural
df_indices.dropna(axis = 0, subset = ['ICMS solidário','Patrimônio Cultural'], inplace = True)
df_indices.isna().sum().sort_values(ascending = False)

Receita Própria           2
Saúde per capita          2
Produção de Alimentos     1
Índice de participação    0
ICMS solidário            0
Cota Mínima               0
Patrimônio Cultural       0
População                 0
Área Geográfica           0
VAF                       0
Municipio                 0
SEF                       0
IBGE                      0
dtype: int64

In [ ]:
# preenchendo os valores ausentes
df_indices.fillna(value = 0.0, inplace = True)
df_indices.isna().sum().sort_values(ascending = False)

Índice de participação    0
ICMS solidário            0
Cota Mínima               0
Receita Própria           0
Saúde per capita          0
Patrimônio Cultural       0
Produção de Alimentos     0
População                 0
Área Geográfica           0
VAF                       0
Municipio                 0
SEF                       0
IBGE                      0
dtype: int64

In [ ]:
# observando esse valores
df_indices[(df_indices['Receita Própria'] == 0.0) |
             (df_indices['Saúde per capita'] == 0.0) |
             (df_indices['Produção de Alimentos'] == 0.0)]

IBGE  SEF  ... ICMS solidário  Índice de participação
2      30    3  ...       0.071239                0.029543
500  4330  433  ...       2.167468                0.996098
667  5733  836  ...       0.045558                0.028070

[3 rows x 13 columns]

In [ ]:
# observando o objeto final
df_indices.head(10)

IBGE  SEF  ... ICMS solidário  Índice de participação
0     10    1  ...       0.037007                0.038541
1     20    2  ...       0.123040                0.072048
2     30    3  ...       0.071239                0.029543
3     40    4  ...       0.021148                0.017551
4     50    5  ...       0.050144                0.026952
5     60    6  ...       0.072727                0.033771
6     70    7  ...       0.010585                0.047961
8     90    9  ...       0.101701                0.035376
9    100   10  ...       0.071689                0.029986
10   110   11  ...       0.133260                0.096369

[10 rows x 13 columns]

In [ ]:
# resetando o index
df_indices.reset_index(drop = True, inplace = True)
df_indices.head(10)

IBGE  SEF  ... ICMS solidário  Índice de participação
0    10    1  ...       0.037007                0.038541
1    20    2  ...       0.123040                0.072048
2    30    3  ...       0.071239                0.029543
3    40    4  ...       0.021148                0.017551
4    50    5  ...       0.050144                0.026952
5    60    6  ...       0.072727                0.033771
6    70    7  ...       0.010585                0.047961
7    90    9  ...       0.101701                0.035376
8   100   10  ...       0.071689                0.029986
9   110   11  ...       0.133260                0.096369

[10 rows x 13 columns]

#### Exportação dos dados processados

In [ ]:
# exportando os dados das alíquotas
df_indices.to_excel('indices.xlsx', index = False)

# exportando os dados do índice mineiro de responsabilidade social
df_imrs.to_excel('imrs_educacao.xlsx')

# exportando os dados do enade
df_enade.to_csv('enade.csv', index = False)

### Análises

Além de trabalhar na construção dos dados é possível utilizar os dataframes do pandas também para fazer análises. Tanto unidimensionais, com análises estatísticas, quanto bidimensionais, com correlações.

É possível também fazer análises de grupos e tabelas cruzadas.

In [ ]:
# observando o objeto imrs
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Conselho de Educação em Atividade
Município                                                                                  ...                                   
Florestal                                                   0.73                           ...                                não
Nova Lima                                                   0.62                           ...                                sim
Juatuba                                                     0.56                           ...                                sim
Jaboticatubas                                               0.69                           ...                                sim
Brumadinho                                                  0.61                           ...                                sim

[5 rows x 73 columns]

In [ ]:
# observando as principais estatísticas
df_imrs.describe()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  % de matrícula em tempo integral_EM
count                                          31.000000                           ...                            31.000000
mean                                            0.533548                           ...                             1.419355
std                                             0.077138                           ...                             2.715038
min                                             0.410000                           ...                             0.000000
25%                                             0.480000                           ...                             0.000000
50%                                             0.520000                           ...                             0.000000
75%                                             0.575000                           ...                             2.160000
max                                             0.730000                           ...                            11.770000

[8 rows x 71 columns]

In [ ]:
# observando as correlações
df_imrs.corr()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  % de matrícula em tempo integral_EM
Índice de Qualidade da Educação para a 4ª série...                                           1.000000                           ...                            -0.036707
Índice de Qualidade da Educação para a 8ª série...                                           0.354040                           ...                             0.169411
Índice de Qualidade da Educação para a 3ª série...                                           0.431142                           ...                             0.176001
Índice de Qualidade da Educação para a 4ª série...                                           0.796777                           ...                            -0.054951
Índice de Qualidade da Educação para a 8ª série...                                           0.354040                           ...                             0.169411
...                                                                                               ...                           ...                                  ...
% de matrícula em tempo integral_Creche                                                     -0.177856                           ...                             0.114163
% de matrícula em tempo integral_Pré-escola                                                  0.233927                           ...                            -0.058849
% de matrícula em tempo integral_EFAI                                                       -0.125611                           ...                            -0.069429
% de matrícula em tempo integral_EFAF                                                       -0.075135                           ...                            -0.179965
% de matrícula em tempo integral_EM                                                         -0.036707                           ...                             1.000000

[71 rows x 71 columns]

In [ ]:
# observando as correlações
variaveis = ['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)',
             'Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática)',
             'Índice de Qualidade da Educação para a 3ª série do EM (matemática)',
             'Índice de Qualidade da Educação para a 4ª série /5º ano do EF (Língua Portuguesa)',
             'Índice de Qualidade da Educação para a 8ª série/9º ano do EF (Língua Portuguesa)',
             'Índice de Qualidade da Educação para a 3ª série do EM (Língua Portuguesa)']

df_imrs[variaveis].corr()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Índice de Qualidade da Educação para a 3ª série do EM (Língua Portuguesa)
Índice de Qualidade da Educação para a 4ª série...                                           1.000000                           ...                                           0.431142                        
Índice de Qualidade da Educação para a 8ª série...                                           0.354040                           ...                                           0.610544                        
Índice de Qualidade da Educação para a 3ª série...                                           0.431142                           ...                                           1.000000                        
Índice de Qualidade da Educação para a 4ª série...                                           0.796777                           ...                                           0.336695                        
Índice de Qualidade da Educação para a 8ª série...                                           0.354040                           ...                                           0.610544                        
Índice de Qualidade da Educação para a 3ª série...                                           0.431142                           ...                                           1.000000                        

[6 rows x 6 columns]

In [ ]:
# observando os dados do enade
df_enade.head()

co_grupo    cd_catadd  ...   co_rs_s11n co_rs_s20n
0  Medicina Veterinária  IES Pública  ...      Auxílio        Sim
1  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
2  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
3  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim
4  Medicina Veterinária  IES Pública  ...  Sem auxílio        Sim

[5 rows x 19 columns]

In [ ]:
# agrupando os dados do enade em relação a instituição
df_enade_group = df_enade.groupby('cd_catadd')

In [ ]:
# tipo do novo objeto
type(df_enade_group)

pandas.core.groupby.generic.DataFrameGroupBy

In [ ]:
# média das variáveis numéricas agrupadas por ies
df_enade_group.mean()

nu_idade    genero   ano_fim_2g
cd_catadd                                    
IES Privada  29.043494  0.734370  2003.432443
IES Pública  26.825166  0.673087  2004.789157

In [ ]:
# média das variáveis numéricas agrupadas por ies
df_enade_group.nunique()

co_grupo  cd_catadd  ...  co_rs_s11n  co_rs_s20n
cd_catadd                         ...                        
IES Privada        17          1  ...           2           2
IES Pública        17          1  ...           2           2

[2 rows x 19 columns]

In [ ]:
# utilizando a função agg (cada linha - IES Pública/IES privada) e lambda (para cada coluna)
df_enade_group.agg(lambda x:x.value_counts().index[0])

co_grupo co_uf_curso  ...   co_rs_s11n  co_rs_s20n
cd_catadd                                ...                         
IES Privada  Serviço Social   São Paulo  ...  Sem auxílio         Sim
IES Pública  Serviço Social   Tocantins  ...  Sem auxílio         Sim

[2 rows x 18 columns]

In [ ]:
# utilizando a função agg e lambda
df_enade_group.agg(lambda x:x.nunique())

co_grupo  co_uf_curso  ...  co_rs_s11n  co_rs_s20n
cd_catadd                           ...                        
IES Privada        17           27  ...           2           2
IES Pública        17           27  ...           2           2

[2 rows x 18 columns]

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], df_enade['cd_catadd'])

cd_catadd        IES Privada  IES Pública
co_regiao_curso                          
Centro-Oeste           17835         4210
Nordeste               26941        12351
Norte                   8680         9260
Sudeste                61455        16250
Sul                    31846         8028

In [ ]:
# testando se o valor é correto
print(df_enade[(df_enade['cd_catadd'] == 'IES Pública') & (df_enade['co_regiao_curso'] == 'Norte')].shape[0])

9260


In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], df_enade['cd_catadd'], margins=True, margins_name="Total")

cd_catadd        IES Privada  IES Pública   Total
co_regiao_curso                                  
Centro-Oeste           17835         4210   22045
Nordeste               26941        12351   39292
Norte                   8680         9260   17940
Sudeste                61455        16250   77705
Sul                    31846         8028   39874
Total                 146757        50099  196856

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], df_enade['cd_catadd'], margins=True, margins_name="Total", normalize=True)

cd_catadd        IES Privada  IES Pública     Total
co_regiao_curso                                    
Centro-Oeste        0.090599     0.021386  0.111985
Nordeste            0.136856     0.062741  0.199598
Norte               0.044093     0.047039  0.091133
Sudeste             0.312183     0.082548  0.394730
Sul                 0.161773     0.040781  0.202554
Total               0.745504     0.254496  1.000000

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], df_enade['cd_catadd'], margins=True, margins_name="Total", normalize='columns')

cd_catadd        IES Privada  IES Pública     Total
co_regiao_curso                                    
Centro-Oeste        0.121527     0.084034  0.111985
Nordeste            0.183576     0.246532  0.199598
Norte               0.059145     0.184834  0.091133
Sudeste             0.418753     0.324358  0.394730
Sul                 0.216998     0.160243  0.202554

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], df_enade['cd_catadd'], margins=True, margins_name="Total", normalize='index')

cd_catadd        IES Privada  IES Pública
co_regiao_curso                          
Centro-Oeste        0.809027     0.190973
Nordeste            0.685661     0.314339
Norte               0.483835     0.516165
Sudeste             0.790876     0.209124
Sul                 0.798666     0.201334
Total               0.745504     0.254496

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], [df_enade['cd_catadd'], df_enade['co_rs_s9n']])

cd_catadd        IES Privada           IES Pública         
co_rs_s9n       Não trabalho Trabalho Não trabalho Trabalho
co_regiao_curso                                            
Centro-Oeste            6301     8543         2683     1181
Nordeste               13751    10020         8473     2648
Norte                   4001     3010         1870     1242
Sudeste                25227    29184         9959     3861
Sul                    10438    17934         5214     2314

In [ ]:
# utilizando o crosstab para região curso e instituição
pd.crosstab(df_enade['co_regiao_curso'], [df_enade['cd_catadd'], df_enade['co_rs_s9n']], normalize='index')

cd_catadd        IES Privada            IES Pública          
co_rs_s9n       Não trabalho  Trabalho Não trabalho  Trabalho
co_regiao_curso                                              
Centro-Oeste        0.336808  0.456650     0.143415  0.063128
Nordeste            0.394102  0.287172     0.242835  0.075891
Norte               0.395239  0.297343     0.184728  0.122691
Sudeste             0.369729  0.427723     0.145960  0.056587
Sul                 0.290752  0.499554     0.145237  0.064457

In [ ]:
# utilizando o crosstab para região curso e instituição - transformando em markdown
pd.crosstab(df_enade['co_regiao_curso'], [df_enade['cd_catadd'], df_enade['co_rs_s9n']], normalize='index').to_markdown()

"| co_regiao_curso   |   ('IES Privada', 'Não trabalho') |   ('IES Privada', 'Trabalho') |   ('IES Pública', 'Não trabalho') |   ('IES Pública', 'Trabalho') |\n|:------------------|----------------------------------:|------------------------------:|----------------------------------:|------------------------------:|\n| Centro-Oeste      |                          0.336808 |                      0.45665  |                          0.143415 |                     0.0631281 |\n| Nordeste          |                          0.394102 |                      0.287172 |                          0.242835 |                     0.0758913 |\n| Norte             |                          0.395239 |                      0.297343 |                          0.184728 |                     0.122691  |\n| Sudeste           |                          0.369729 |                      0.427723 |                          0.14596  |                     0.0565872 |\n| Sul               |                        

**Tabela**

| co_regiao_curso   |   ('IES Privada', 'Não trabalho') |   ('IES Privada', 'Trabalho') |   ('IES Pública', 'Não trabalho') |   ('IES Pública', 'Trabalho') |
|:------------------|----------------------------------:|------------------------------:|----------------------------------:|------------------------------:|
| Centro-Oeste      |                          0.336808 |                      0.45665  |                          0.143415 |                     0.0631281 |
| Nordeste          |                          0.394102 |                      0.287172 |                          0.242835 |                     0.0758913 |
| Norte             |                          0.395239 |                      0.297343 |                          0.184728 |                     0.122691  |
| Sudeste           |                          0.369729 |                      0.427723 |                          0.14596  |                     0.0565872 |
| Sul               |                          0.290752 |                      0.499554 |                          0.145237 |                     0.0644568 |

---

### Combinação de dataframes

Uma outra funcionalidade dos dataframes do pandas é combinar diferentes conjunto de dados. Por exemplo, temos os dados do índice mineiro de responsabilidade social por município da RM de Belo Horizonte, mas também temos os índices da lei Robin Hood para os municípios mineiros.

Ambos tem a mesma informação, que são os municípios. Nesse caso é possível agrupar os dois dataframes sob a mesma variável.

In [ ]:
# observando o objeto imrs
df_imrs.head()

Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)  ...  Conselho de Educação em Atividade
Município                                                                                  ...                                   
Florestal                                                   0.73                           ...                                não
Nova Lima                                                   0.62                           ...                                sim
Juatuba                                                     0.56                           ...                                sim
Jaboticatubas                                               0.69                           ...                                sim
Brumadinho                                                  0.61                           ...                                sim

[5 rows x 73 columns]

# Curso de Introdução ao Python: Módulo 4

---

<center>
<img src="https://github.com/marcelorandolfo/data-science/raw/master/images/imagem_curso.png" width="3000"/>
</center>


**Marcelo Randolfo** \\
**Fundação João Pinheiro**

In [ ]:
# definindo um novo dataframe com algumas variáveis
variaveis_matematica = ['Índice de Qualidade da Educação para a 4ª série /5º ano do EF (matemática)',
       'Índice de Qualidade da Educação para a 8ª série/9º ano do EF (matemática)',
       'Índice de Qualidade da Educação para a 3ª série do EM (matemática)']

df_1 = df_imrs[variaveis_matematica].reset_index()


df_1.head()

Município  ...  Índice de Qualidade da Educação para a 3ª série do EM (matemática)
0      Florestal  ...                                               0.48                 
1      Nova Lima  ...                                               0.37                 
2        Juatuba  ...                                               0.40                 
3  Jaboticatubas  ...                                               0.33                 
4     Brumadinho  ...                                               0.34                 

[5 rows x 4 columns]

In [ ]:
# definindo um novo dataframe com algumas variáveis
variaveis_indices = ['Municipio', 'Área Geográfica','População']

df_2 = df_indices[variaveis_indices]
df_2.head()

Municipio  Área Geográfica  População
0  Abadia dos Dourados         0.150510   0.033016
1               Abaeté         0.309244   0.109770
2           Abre Campo         0.080162   0.063556
3              Acaiaca         0.017429   0.018867
4              Açucena         0.138312   0.044736

<img src = 'https://lh3.googleusercontent.com/proxy/PxYOagmmNi3lvOTUWOfe_9_XZRm54NW8i1L_mXwpB-dy5MDAgEvitgr3AJ_qvyqOug44jUaqF_RCmJDhmSxfBnhVgEZhraxMS0gHf03TiV1k_zapU31SreIqPALY-DKh5RBs1A8U-m5gw62K48uxUz8QjyC-B2ym8uasog'>

In [ ]:
# combinando dois dataframes
df_combinado = pd.merge(df_1, df_2,how='inner', left_on='Município', right_on = 'Municipio')
df_combinado

Município  ...  População
0              Florestal  ...   0.035245
1          Jaboticatubas  ...   0.095154
2       São José da Lapa  ...   0.112269
3                Igarapé  ...   0.203342
4                 Baldim  ...   0.036970
5                Raposos  ...   0.077255
6                  Caeté  ...   0.211245
7   São Joaquim de Bicas  ...   0.149172
8            Lagoa Santa  ...   0.304821
9         Belo Horizonte  ...  11.866856
10            Matozinhos  ...   0.178659
11        Pedro Leopoldo  ...   0.303551
12    Taquaraçu de Minas  ...   0.019259
13            Vespasiano  ...   0.602779
14           Santa Luzia  ...   1.035175
15            Nova União  ...   0.027045
16              Contagem  ...   3.136008
17               Ibirité  ...   0.851272
18            Esmeraldas  ...   0.333283
19          Capim Branco  ...   0.046077
20               Confins  ...   0.031792
21                Sabará  ...   0.644080
22          Mário Campos  ...   0.072824
23    Ribeirão das Neves  ...   1.581848

[24 rows x 7 columns]

In [ ]:
# testando o a combinação
df_indices[['Área Geográfica','População']][df_indices['Municipio'] == 'Ribeirão das Neves']

Área Geográfica  População
556         0.026359   1.581848

In [ ]:
# combinando os dataframes pelo outer
df_combinado_outer = pd.merge(df_1, df_2,how='outer', left_on='Município', right_on = 'Municipio')
df_combinado_outer

Município  ...  População
0        Florestal  ...   0.035245
1        Nova Lima  ...        NaN
2          Juatuba  ...        NaN
3    Jaboticatubas  ...   0.095154
4       Brumadinho  ...        NaN
..             ...  ...        ...
749            NaN  ...   0.049649
750            NaN  ...   0.025415
751            NaN  ...   0.201070
752            NaN  ...   0.024810
753            NaN  ...   0.012055

[754 rows x 7 columns]

In [ ]:
# combinando os dataframes pelo left
df_combinado_left = pd.merge(df_1, df_2,how='left', left_on='Município', right_on = 'Municipio')
df_combinado_left

Município  ...  População
0              Florestal  ...   0.035245
1              Nova Lima  ...        NaN
2                Juatuba  ...        NaN
3          Jaboticatubas  ...   0.095154
4             Brumadinho  ...        NaN
5                Sarzedo  ...        NaN
6             Itatiaiuçu  ...        NaN
7       São José da Lapa  ...   0.112269
8                Igarapé  ...   0.203342
9                 Baldim  ...   0.036970
10               Raposos  ...   0.077255
11                 Caeté  ...   0.211245
12  São Joaquim de Bicas  ...   0.149172
13                 Betim  ...        NaN
14           Lagoa Santa  ...   0.304821
15             Rio Acima  ...        NaN
16        Belo Horizonte  ...  11.866856
17            Matozinhos  ...   0.178659
18        Pedro Leopoldo  ...   0.303551
19    Taquaraçu de Minas  ...   0.019259
20            Vespasiano  ...   0.602779
21           Santa Luzia  ...   1.035175
22            Nova União  ...   0.027045
23              Contagem  ...   3.136008
24               Ibirité  ...   0.851272
25            Esmeraldas  ...   0.333283
26          Capim Branco  ...   0.046077
27               Confins  ...   0.031792
28                Sabará  ...   0.644080
29          Mário Campos  ...   0.072824
30    Ribeirão das Neves  ...   1.581848

[31 rows x 7 columns]

<img src  = 'http://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png'>

In [ ]:
# combinando os dataframes pelo right
df_combinado_right = pd.merge(df_1, df_2,how='right', left_on='Município', right_on = 'Municipio')
df_combinado_right

Município  ...  População
0           Florestal  ...   0.035245
1       Jaboticatubas  ...   0.095154
2    São José da Lapa  ...   0.112269
3             Igarapé  ...   0.203342
4              Baldim  ...   0.036970
..                ...  ...        ...
742               NaN  ...   0.049649
743               NaN  ...   0.025415
744               NaN  ...   0.201070
745               NaN  ...   0.024810
746               NaN  ...   0.012055

[747 rows x 7 columns]